### DISCUSSION TITLE
AttributeError: 'spacy.morphology.Morphology' object has no attribute 'tag_map'

### DISCUSSION THREAD

**Discussion by pradeepdev-1995** [explosion/spaCy#10890] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour
When I am trying the below code
```
import en_core_web_sm
nlp = en_core_web_sm.load()
doc = nlp("my sentence")
for token in doc:
         morph = nlp.vocab.morphology.tag_map[token.tag_]
         print(morph)
```
I am getting the following spacy error

```AttributeError: 'spacy.morphology.Morphology' object has no attribute 'tag_map'```

## Your Environment

* Operating System:  Ubuntu 18.04.6
* Python Version Used: 3.9.13
* spaCy Version Used:
spacy==3.2.4
spacy-legacy==3.0.9
spacy-loggers==1.0.2
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm- 3.2.0/en_core_web_sm-3.2.0-py3-none-any.whl

**Comment by polm** [comment, marked as answer]
What are you trying to do exactly? 

I believe the code you provide would work in v2, but it doesn't work in v3 - [Morphology](https://spacy.io/api/morphology) objects do not have a `tag_map`. 

Also see [this v2 migration note](https://spacy.io/usage/v3/#migrating-training-mappings-exceptions).

**Reply by pradeepdev-1995** [reply]
@polm 
Thanks for your reply.
The below-mentioned code
```
import en_core_web_sm
nlp = en_core_web_sm.load()
doc = nlp("this is sample sentence for testing")
for token in doc:
  morph = nlp.vocab.morphology.tag_map[token.tag_]
  print(morph)
```
outputs the following
```
{74: 90}
{74: 100, 'Number_sing': True, 'Person_three': True, 'Tense_pres': True, 'VerbForm_fin': True}
{74: 92, 'Number_sing': True}
{74: 92, 'Number_sing': True}
{74: 85}
{74: 92, 'Number_sing': True}
```
in ```spacy==2.2.1```.
If i need the same output with spacy v3( say ```spacy==3.1.6```)
then how should I change the above code in the Spacy v3 scenario?

**Reply by polm** [reply]
You can get that information from `tok.morph.to_dict()`. The format is not exactly the same, it's more straightforward.

```
import en_core_web_sm

nlp = en_core_web_sm.load()
doc = nlp("this is sample sentence for testing")
for token in doc:
    print(token.morph.to_dict())
```

```
{'Number': 'Sing', 'PronType': 'Dem'}
{'Mood': 'Ind', 'Number': 'Sing', 'Person': '3', 'Tense': 'Pres', 'VerbForm': 'Fin'}
{'Number': 'Sing'}
{'Number': 'Sing'}
{}
{'Number': 'Sing'}
```

**Reply by pradeepdev-1995** [reply]
@polm  Thank you. It worked for me
